mava/networks/utils/sable/discrete_trainer_executor.py

```diff
@@ -0,0 +1,337 @@
+    )
+
+    # Apply the encoder
+    v_loc, obs_rep, _ = execute_encoder_parallel(
+        encoder=encoder, obs=obs, decayed_hstate=hstates[0], timestep_id=timestep_id
+    )
+
+    # Apply the decoder
+    _ = autoregressive_act(
+        decoder=decoder,
+        obs_rep=obs_rep,
+        legal_actions=legal_actions,
+        hstates=hstates[1],
+        timestep_id=timestep_id,
+        key=key,
+    )
+
+    return v_loc
+
+
+def execute_encoder_parallel(
+    encoder: nn.Module,
+    obs: chex.Array,
+    decayed_hstate: chex.Array,
+    timestep_id: chex.Array,
+) -> Tuple[chex.Array, chex.Array, chex.Array]:
+    """Parallel encoding for discrete action spaces."""
+    # Apply the encoder
+    v_loc, obs_rep, updated_hstate = encoder.recurrent(obs, decayed_hstate, timestep_id)
+    return v_loc, obs_rep, updated_hstate
+
+
+def execute_encoder_chunkwise(
+    encoder: nn.Module,
+    obs: chex.Array,
+    decayed_hstate: chex.Array,
+    timestep_id: chex.Array,
+    chunk_size: int,
+) -> Tuple[chex.Array, chex.Array, chex.Array]:
+    """Chunkwise encoding for Non-memory Sable and for discrete action spaces."""
+    # Get the batch and sequence dimensions
+    batch_dim, agents_per_seq = obs.shape[:2]
+    # Initialize the value location and observation representation
+    v_loc = jnp.zeros((batch_dim, agents_per_seq, 1))
+    obs_rep = jnp.zeros((batch_dim, agents_per_seq, encoder.net_config.embed_dim))
+    # Apply the encoder per chunk
+    num_chunks = agents_per_seq // chunk_size
+    for chunk_id in range(0, num_chunks):
+        start_idx = chunk_id * chunk_size
+        end_idx = (chunk_id + 1) * chunk_size
+        chunk_obs = obs[:, start_idx:end_idx]
+        chunk_timestep_id = timestep_id[:, start_idx:end_idx]
+        # Apply parallel encoding per chunk
+        chunk_v_loc, chunk_obs_rep, decayed_hstate = execute_encoder_parallel(
+            encoder=encoder,
+            obs=chunk_obs,
+            decayed_hstate=decayed_hstate,
+            timestep_id=chunk_timestep_id,
+        )
+        v_loc = v_loc.at[:, start_idx:end_idx].set(chunk_v_loc)
+        obs_rep = obs_rep.at[:, start_idx:end_idx].set(chunk_obs_rep)
+
+    return v_loc, obs_rep, decayed_hstate
+
+
+def autoregressive_act(
+    decoder: nn.Module,
+    obs_rep: chex.Array,
+    hstates: chex.Array,
+    legal_actions: chex.Array,
+    timestep_id: chex.Array,
+    key: chex.PRNGKey,
+) -> Tuple[chex.Array, chex.Array, chex.Array]:
+    # Get the batch size, sequence length, and action dimension
+    batch_size, n_agents, action_dim = legal_actions.shape
+
+    # Create a shifted action sequence for predicting the next action
+    # Initialize the shifted action sequence.
+    shifted_actions = jnp.zeros((batch_size, n_agents, action_dim + 1))
+    # Set the start-of-timestep token (first action as a "start" signal)
+    shifted_actions = shifted_actions.at[:, 0, 0].set(1)
+
+    # Define the output action and output action log sizes
```

Comment:
Same here for shapes and unneeded comments
```suggestion
    batch_size, n_agents, action_dim = legal_actions.shape

    # Create a shifted action sequence and set the start-of-timestep token
    shifted_actions = jnp.zeros((batch_size, n_agents, action_dim + 1))
    shifted_actions = shifted_actions.at[:, 0, 0].set(1)

```